### PR TITLE
feat: display image message in conversation [AR-1150]

### PR DESCRIPTION
**Pull request by gongracr**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/AR-1150" title="AR-1150" target="_blank">SubtaskAR-1150</a>  Implement preview behavior for displaying images vs files on conversation
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

Adaptation on the Android client of the ability of downloading, decrypting and displaying images assets from Kalium. The `ImageMessage` object now receives a ByteArray of decrypted data to be displayed. 

Needs releases with:
- https://github.com/wireapp/kalium/pull/377

- [x] GitHub link to other pull request

### How to test
Check the attached video

### Attachments (Optional)
https://user-images.githubusercontent.com/2468164/162405564-2746c94c-3f6f-4d18-a875-6ca23367cd26.mp4

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [x] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [x] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/model/MessageViewWrapper.kt
Comment (trOnk12) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/model/MessageViewWrapper.kt:20]: confusing name to me a bit, what does it wrap, why would it wrap anything ?

**Comment by AndroidBob** [comment]
Build [1](https://10.10.124.134/job/wire-android-reloaded/job/PR-493/1/) **failed**.

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/ConversationViewModel.kt
Comment (trOnk12) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/ConversationViewModel.kt:247]: I saw this ByteArray(16), why are we returning it actually ? and why not ByteArray(15) ? :P just confusing why it is here, maybe a comment would do with a explanation

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/model/MessageViewWrapper.kt
Comment (gongracr) [reply app/src/main/kotlin/com/wire/android/ui/home/conversations/model/MessageViewWrapper.kt:20]: I added wrapper since it also contains info like `messageSource` that is not directly related with the UI. I can also rename to `messageUI` if preferred

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/ConversationViewModel.kt
Comment (gongracr) [reply app/src/main/kotlin/com/wire/android/ui/home/conversations/ConversationViewModel.kt:208]: replied above

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/ConversationViewModel.kt
Comment (gongracr) [reply app/src/main/kotlin/com/wire/android/ui/home/conversations/ConversationViewModel.kt:247]: I return for now a default byte array in case there was an error downloading the asset data. Later on, a default placeholder will be shown instead inside the `MessageImage` object since the call to `rawImgData.toBitmap()` will fail on an empty byte array. 
About the 16, it's just a convention for byte array initialisation. 
I will add a comment for now.

**Comment by AndroidBob** [comment]
Build [2](https://10.10.124.134/job/wire-android-reloaded/job/PR-493/2/) **failed**.

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/ConversationViewModel.kt
Comment (trOnk12) [reply app/src/main/kotlin/com/wire/android/ui/home/conversations/ConversationViewModel.kt]: if you just add a comment what you wrote above, it would be aweasome, because it explains it all nicely

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/MessageItem.kt
Comment (trOnk12) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/MessageItem.kt:139]: how about creating some sort of data class for example that would hold this values and also have transform method to do what you did here : 

```
    val width = MaterialTheme.wireDimensions.messageImageMaxWidth
    val height = width.value * realImgHeight.toFloat() / realImgWidth
    ```

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/model/MessageViewWrapper.kt
Comment (trOnk12) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/model/MessageViewWrapper.kt:31]: perhpas this method could have a function to transofrm the size of the image

**Review by trOnk12 — COMMENTED** [review]
great job Gonzo 👍, just some small comments about perhaps renaming the classes and moving some stuff,  amazing to see how Wire begins to take a shape of a real product ❤️

**Comment by AndroidBob** [comment]
Build [3](https://10.10.124.134/job/wire-android-reloaded/job/PR-493/3/) **failed**.

**Comment by AndroidBob** [comment]
Build [5](https://10.10.124.134/job/wire-android-reloaded/job/PR-493/5/) **failed**.

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/MessageItem.kt
Comment (gongracr) [reply app/src/main/kotlin/com/wire/android/ui/home/conversations/MessageItem.kt]: I created a wrapper class called `ImageMessageParams` containing these normalized width and height values

**Comment by AndroidBob** [comment]
Build [6](https://10.10.124.134/job/wire-android-reloaded/job/PR-493/6/) **failed**.

**Review by trOnk12 — APPROVED** [review]
awesome job, looking great 🥇 

looks good to me, maybe someone else wants to take a look before merge, but on my side, all is looking clean 🍰

**Comment by AndroidBob** [comment]
Build [7](https://10.10.124.134/job/wire-android-reloaded/job/PR-493/7/) **failed**.

**Comment by AndroidBob** [comment]
Build [8](https://10.10.124.134/job/wire-android-reloaded/job/PR-493/8/) **failed**.

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/ConversationViewModel.kt
Comment (MohammedMokresh) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/ConversationViewModel.kt:258]: maybe we can have enum for the mimetypes , so it can be reused for other cases later ?

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/ConversationViewModel.kt
Comment (MohammedMokresh) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/ConversationViewModel.kt:259]: can we add those strings to the strings.xml file

**Review by MohammedMokresh — APPROVED** [review]

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/ConversationViewModel.kt
Comment (gongracr) [reply app/src/main/kotlin/com/wire/android/ui/home/conversations/ConversationViewModel.kt:259]: This is not final. We need to define an error UI to handle this, so I would avoid to include it to the strings file so that it doesn't create unneeded translations. I added a TODO comment instead

**Comment by AndroidBob** [comment]
Build [9](https://10.10.124.134/job/wire-android-reloaded/job/PR-493/9/) **failed**.

**Comment by AndroidBob** [comment]
Build [10](https://10.10.124.134/job/wire-android-reloaded/job/PR-493/10/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv0.0.1(3071499)(3071499)-dev-release.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/dev/release/com.wire.android-vv0.0.1(3071499)(3071499)-dev-release.apk)

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/model/MessageViewWrapper.kt
Comment (yamilmedina) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/model/MessageViewWrapper.kt:42]: Just out of curiosity, the `equals` and `hashCode` methods generally when auto-generating, it will consider all the fields values on the object for the comparison, I think to represent real state and diff between instances. Did we get intentionally out of that comparison, the `width` and `height` ? Which would kind of make sense I think

**Review by yamilmedina — APPROVED** [review]
Great work as always 🍦 !  Just left a question there to know if that behavior it's intentional or not :D

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/model/MessageViewWrapper.kt
Comment (gongracr) [reply app/src/main/kotlin/com/wire/android/ui/home/conversations/model/MessageViewWrapper.kt:42]: Yes, indeed that shouldn't be a problem. The `equals` comparison is basically to allow us to compare the raw content of the image. The `width` and `height` are just metadata that we use to pass through the final `MessageImage` object used to display the inline image in the conversation. So basically, 2 `ImageMessage` objects with the same raw image but different `width` and `height` (although I'd say in practise that would be impossible) would still be considered the same image. 
But good point, we should definitely need to take a look into this the moment we start considering other new features like message forwarding from different conversations, etc...

**Pull request merged**
